cleverhans/model_abs.py

```diff
@@ -0,0 +1,73 @@
+from abc import ABCMeta
+from keras.models import Model
+
+
+class ModelAbstraction:
+    """
+    An abstract interface for a model that exposes the required functionalities
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
+        """
+        :param model: A function that takes a symbolic input and returns the
+                      symbolic output for the model's predictions.
+        """
+        raise NotImplementedError('Model abstraction not implementated.')
+
+    def get_hidden(self, layer):
+        """
+        :return: A symbolic representation of the hidden features
+        :param layer: The name of the hidden layer to return features at.
+        """
+        error = 'Feature extraction for hidden layers not implemented'
+        raise NotImplementedError(error)
+
+    def get_logits(self):
```

Comment:
Here `get_hidden` depends on the model-specific layer names. `get_logits` and `get_probs` are supposed to be independent of the layer names. Possibly, `get_logits` and `get_probs` can use `get_hidden` but after they find the appropriate layer. Your suggestion can be done but it would be by having multiple special cases inside `get_hidden` that would complicate the method. 

`get_logits` and `get_probs` are special cases because they are currently the only methods needed by the core code. Here are the usages:
https://github.com/tensorflow/cleverhans/blob/fa17e16051f60221926b7fb884e352c15e484134/cleverhans/utils_tf.py#L52-L56
https://github.com/tensorflow/cleverhans/blob/fa17e16051f60221926b7fb884e352c15e484134/cleverhans/utils_tf.py#L213-L219

In the future, we might want to have something like `get_embed` for the final feature representation before logits, eg. FC7.